zebra-chain/src/mmr.rs

```diff
@@ -0,0 +1,190 @@
+}
+
+/// History tree structure.
+pub struct HistoryTree {
+    network: Network,
+    network_upgrade: NetworkUpgrade,
+    // Merkle mountain range tree.
+    // This is a "runtime" structure used to add / remove nodes, and it's not
+    // persistent.
+    inner: Tree,
+    // The number of nodes in the tree.
+    size: u32,
+    // The peaks of the tree, indexed by their position in the array representation
+    // of the tree. This can be persisted to save the tree.
+    // TODO: use NodeData instead of Entry to save space? Requires re-deriving
+    // the entry metadata from its position.
+    peaks: HashMap<u32, Entry>,
+}
+
+impl HistoryTree {
+    /// Create a new history tree with a single block.
+    pub fn new_from_block(
+        network: Network,
+        block: Arc<Block>,
+        sapling_root: &sapling::tree::Root,
+        _orchard_root: Option<&orchard::tree::Root>,
+    ) -> Result<Self, io::Error> {
+        let height = block
+            .coinbase_height()
+            .expect("block must have coinbase height during contextual verification");
+        let network_upgrade = NetworkUpgrade::current(network, height);
+        // TODO: handle Orchard root
+        let (tree, entry) = Tree::new_from_block(network, block, sapling_root)?;
+        let mut peaks = HashMap::new();
+        peaks.insert(0u32, entry);
+        Ok(HistoryTree {
+            network,
+            network_upgrade,
+            inner: tree,
+            size: 1,
+            peaks,
+        })
+    }
+
+    /// Add block data to the tree.
+    pub fn push(
+        &mut self,
+        block: Arc<Block>,
+        sapling_root: &sapling::tree::Root,
+        _orchard_root: Option<&orchard::tree::Root>,
+    ) -> Result<(), HistoryTreeError> {
+        // TODO: handle orchard root
+        let new_entries = self
+            .inner
+            .append_leaf(block, sapling_root)
+            .map_err(|e| HistoryTreeError::InnerError { inner: e })?;
+        for entry in new_entries {
+            // Not every entry is a peak; those will be trimmed later
+            self.peaks.insert(self.size, entry);
+            self.size += 1;
+        }
+        // TODO: should we really prune every iteration?
+        self.prune()?;
+        // TODO: implement network upgrade logic: drop previous history, start new history
+        Ok(())
+    }
+
+    /// Extend the history tree with the given blocks.
+    pub fn extend<
+        'a,
+        T: IntoIterator<
+            Item = (
+                Arc<Block>,
+                &'a sapling::tree::Root,
+                Option<&'a orchard::tree::Root>,
+            ),
+        >,
+    >(
+        &mut self,
+        iter: T,
+    ) -> Result<(), HistoryTreeError> {
+        for (block, sapling_root, orchard_root) in iter {
+            self.push(block, sapling_root, orchard_root)?;
+        }
+        Ok(())
+    }
+
+    /// Prune tree, removing all non-peak entries.
+    fn prune(&mut self) -> Result<(), io::Error> {
```

Comment:
Testing: Since this is non-trivial code, let's add a proptest for this function.

I think the `PreparedChain` proptest strategy produces the kind of data you want here. You can use the generated `count` to split the chain, then prune it back to that block. (`count` lets proptest automatically minimise test cases.)

https://github.com/ZcashFoundation/zebra/blob/0b611eb77056242694f4d7fe72b78a216e7cca59/zebra-state/src/service/non_finalized_state/tests/prop.rs#L17-L27

If the tests don't work with the default`PreparedChain` genesis/NU5 strategy, we can add a constructor which takes a ledger state. Then we can use `prop_flat_map` to feed random ledger states to `PreparedChain`:
https://github.com/ZcashFoundation/zebra/blob/9688811e35c466307080e8abccc8485168621bfc/zebra-chain/src/block/arbitrary.rs#L170-L179

Reply:
It looks like `forked_equals_pushed` does all these tests for you. We might just need to change its initial `LegderState` strategy.